### PR TITLE
Providers refactor

### PR DESCRIPTION
**Pull request by jkrocil**
Added basis for container providers addition

Added containers top container menu branches

{{pytest: cfme/tests/cloud/test_providers.py cfme/tests/infrastructure/test_providers.py -v -k provider_crud --use-provider default --long-running }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 8706df2a8228f97900d12e33ac0e102858986494:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/provider.py:25:1: F401 'ensure_browser_open' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f0ae4e065d0586593a1f5f0680b4be36e27b851f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 749851c7c13980010323e95a7d6a1b39a3a0c57d:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: cfme/common/provider.py [hunk 1]
```diff
@@ -75,6 +88,7 @@ def __init__(self, **kwargs):
             self.candu = kwargs.get('candu')
             self.domain = kwargs.get('domain')
             self.ssh = kwargs.get('ssh')
+            self.token = kwargs.get('token')
```
Comment (psav) [inline cfme/common/provider.py]: As we're in here, I've always hated this, do you think we could make it a ```credential.type``` attribute?

Comment (jkrocil) [inline cfme/common/provider.py]: Absolutely. I will nod at and agree with every comment here as I want it merged.

**Comment by dajoRH** [comment]
Lint report for commit a2b39da03aa67c928224eb3878f1d5af9a8ab65d:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by jkrocil** [comment]
Moved `num_vm` and `num_template` to `CloudInfraProvider`, also moved policy assignment related stuff there and removed `_load_details` as its not necessary anymore.
Also replaced any `sel.force_navigate('{page_name}_provider'...)` call that would lead to detail page by `self.load_details()` call that also checks if we are already there.

**Comment by jkrocil** [comment]
And `wait_for_creds_ok` got moved as well because container providers currently do not have a way to let us know if their creds are valid. And since `BaseProvider` is a basis for that type as well it had to be moved.

**Comment by jkrocil** [comment]
@psav re your `num_vm`/`num_template` being used with container providers - we fetch the data the same way for infra and cloud, from the `vms` table. So we can have a `num_vm` in place for both and just point `num_instance` at it and were done. We cant use the same method for `num_pod` or `num_image` as its in a separate table.
What we could do is use the `BaseProvider.category` to decide from where to fetch it but... do we really want to have a `BaseProvider` with `.num_vm`, `.num_template`, `.num_instance`, `.num_image`, `.num_pod`, `.num_container_image` and possibly more in the future?

**Comment by jkrocil** [comment]
@mfalesni ^

**Comment by mfalesni** [comment]
I'd personally say that having so many methods in the base class woiuld not be the best idea. For us, the main reason for sharing the common root is DRY. If you work with some area, you always know what objects you use so you will not call `num_vm` at the container objects and vice versa.

**Comment by psav** [comment]
It was only my intention to keep methods in there that were common across all, but it seems they are not, so split it out ;)

**Comment by mfalesni** [comment]
muhaha

**Pull request merged**